site/src/routes/pins-add.js

```diff
@@ -3,43 +3,114 @@ import { verifyToken } from '../utils/utils.js'
 import * as PinataPSA from '../pinata-psa.js'
 import { JSONResponse } from '../utils/json-response.js'
 import * as nfts from '../models/nfts.js'
+import * as cluster from '../cluster.js'
 
 /**
  * @param {FetchEvent} event
  * @returns {Promise<JSONResponse>}
  */
-export async function pinsAdd (event) {
+export async function pinsAdd(event) {
   const result = await verifyToken(event)
   if (!result.ok) {
     return HTTPError.respond(result.error)
   }
   const { user, tokenName } = result
   const pinData = await event.request.json()
-  pinData.meta = pinData.meta || {}
-  pinData.meta.sub = user.sub
 
-  const res = await PinataPSA.pinsAdd(pinData)
-  if (!res.ok) {
-    return PinataPSA.parseErrorResponse(res.error)
+  if (typeof pinData.cid !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid CID' } },
+      { status: 400 }
+    )
+  }
+  if (pinData.name && typeof pinData.name !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid name' } },
+      { status: 400 }
+    )
+  }
+
+  const name = pinData.name
+
+  let meta
+  if (pinData.meta) {
+    if (typeof pinData.meta !== 'object' || Array.isArray(pinData.meta)) {
+      return new JSONResponse(
+        { error: { reason: 'INVALID_PIN_DATA', details: 'invalid metadata' } },
+        { status: 400 }
+      )
+    }
+    meta = Object.fromEntries(
+      Object.entries(pinData.meta).filter(([, v]) => typeof v === 'string')
+    )
   }
```

Comment:
can't we just try catch JSON.stringify ? plus kv metadata can only be 1024 bytes in length

Reply:
The pin `meta` MUST be an object and it MUST be a `Record<string, string>`. Any values that are not strings are ignored and not persisted. This is what the spec says.

This is just checking that the passed `meta` is a plain object and then filtering out any values that are not, in order to conform to the spec.

My gut instinct is that the majority of people will not encounter the 1024 bytes limit. If we find lots of people are seeing it then we'll prioritise moving to a new solution for listing pins sooner (the pin metadata is in the KV metadata so we can filter by it in the pin listing endpoint withough having to fetch every single NFT in the KV).